addons/l10n_in_ewaybill_stock/models/l10n_in_ewaybill.py

```diff
@@ -0,0 +1,603 @@
+        if partner.country_id.code == "IN":
+            if partner.state_id and not partner.state_id.l10n_in_tin:
+                message.append(_("- TIN number not set in state %s", partner.state_id.name))
+            if not partner.state_id:
+                message.append(_("- State is required"))
+            if not partner.zip or not re.match("^[0-9]{6}$", partner.zip):
+                message.append(_("- Zip code required and should be 6 digits"))
+        elif not partner.country_id:
+            message.append(_("- Country is required"))
+        if message:
+            message.insert(0, "%s" % partner.display_name)
+        return message
+
+    def _check_document_number(self):
+        if not re.match("^.{1,16}$", self.document_number):
+            return [_("Document number should be set and not more than 16 characters")]
+        return []
+
+    def _check_lines(self):
+        error_message = []
+        AccountEDI = self.env['account.edi.format']
+        for line in self.move_ids:
+            if not (hsn_code := AccountEDI._l10n_in_edi_extract_digits(line.product_id.l10n_in_hsn_code)):
+                error_message.append(_("HSN code is not set in product %s", line.product_id.name))
+            elif not re.match("^[0-9]+$", hsn_code):
+                error_message.append(_(
+                    "Invalid HSN Code (%s) in product %s", hsn_code, line.product_id.name
+                ))
+        return error_message
+
+    def _check_gst_treatment(self):
+        partner, gst_treatment = self._get_gst_treatment()
+        if not gst_treatment:
+            return [_("Set GST Treatment for in %s", partner.display_name)]
+        return []
+
+    def _get_gst_treatment(self):
+        if self.picking_type_code == 'incoming':
+            partner = self.partner_bill_from_id
+        else:
+            partner = self.partner_bill_to_id
+        return partner, partner.l10n_in_gst_treatment
+
+    def _write_error(self, error_message, blocking_level='error'):
+        self.write({
+            'error_message': error_message,
+            'blocking_level': blocking_level,
+        })
+
+    def _write_successfully_response(self, response_vals):
+        response_vals.update({
+            'error_message': False,
+            'blocking_level': False,
+        })
+        self.write(response_vals)
+
+    def _handle_internal_warning_if_present(self, response):
+        if warnings := response.get('odoo_warning'):
+            for warning in warnings:
+                if warning.get('message_post'):
+                    odoobot = self.env.ref("base.partner_root")
+                    self.message_post(
+                        author_id=odoobot.id,
+                        body=warning.get('message')
+                    )
+                else:
+                    self._write_error(warning.get('message'))
+
+    def handle_error_if_present(self, ewaybill_error):
+        if is_error := isinstance(ewaybill_error, EWayBillError):
```

Comment:
We are still mixing things up here.  The method in tools/ewaybill_api.py should raise if it is an error, not return it as such.  In the except of the try..except, we can then pass this ewaybill error to this method.   e.g. the cancel in L395 should be surrounded by a try..except.

Reply:
Indeed